torch/onnx/_internal/fx/passes/functionalization.py

```diff
@@ -0,0 +1,104 @@
+from __future__ import annotations
+
+import torch
+import torch.func
+import torch.fx
+
+from torch.fx.experimental import proxy_tensor
+from torch.onnx._internal import _beartype
+from torch.onnx._internal.fx import _pass
+from torch.onnx._internal.fx.passes import _utils
+
+
+class Functionalize(_pass.Transform):
+    """Functionalize a GraphModule.
+
+    This pass utilizes ``torch.func.functionalize`` to convert a GraphModule into a
+    functional form. The two main functionalities are (copied from its documentations):
+
+    * ``torch.func.functionalize`` removes (intermediate) mutations and aliasing from a
+    function, while preserving the function's semantics.
+
+    * ``torch.func.functionalize`` also removes mutations (and views) that were performed
+    on function inputs. However to preserve semantics, functionalize will "fix up" the
+    mutations after the transform has finished running, by detecting if any tensor inputs
+    "should have" been mutated, and copying the new data back to the inputs if necessary.
+    For example, consider::
+
+        def fn(a, b):
+            a.add_(b)
+            return a
+
+      For a call like `fn(x, y)`, the variable `x` outside is also mutated. Hence just
+      functionalizing is not enough for preserving the original semantics. A "special"
+      input mutation step needs to be inserted at the end.::
+
+        # After functionalization, without input mutation "fix up".
+        # This is not semantically the same. The variable outside the function call that
+        # was passed in as `a` is not mutated.
+        def fn(a, b):
+            new_a = a + b
+            return new_a
+
+        # Functionalization with input mutation "fix up" that preserves semantics.
+        def fn(a, b):
+            new_a = a + b
+
+            # Copying the new data back to the inputs
+            a.copy_(new_a)
+
+            return new_a
+
+    For ONNX inference, it is recommended to run ``RemoveInputMutation`` after this pass.
+    ``RemoveInputMutation`` removes the "fix up" nodes that were added by ``Functionalize``,
+    which are not needed for ONNX inference.
```

Comment:
Why not remove fix up in this pass?

Reply:
Because input mutation might be useful in training.

Reply:
So we are likely to bypass the RemoveInputMutation pass if model is in training mode?

Reply:
Yes.

To add more details, I think how to make use of this information depends on how `dynamo_export` integrates with training. But generally speaking, it is always better to preserve as much information as possible.